dot/parachain/prospective-parachains/active_leaves_update.go

```diff
@@ -0,0 +1,383 @@
+			ancestryBlockInfo[idx] = relayChainBlockInfo{
+				Hash:        a.Hash,
+				Number:      a.Number,
+				StorageRoot: a.StorageRoot,
+			}
+			ancestorsHashes[idx] = a.Hash.String()
+		}
+
+		var prevFragmentChains map[parachaintypes.ParaID]*fragmentChain
+		if len(ancestry) > 0 {
+			prevFragmentChains = pp.view.perRelayParent[ancestry[0].Hash].fragmentChains
+		}
+
+		fragmentChains := make(map[parachaintypes.ParaID]*fragmentChain)
+		for paraID, claimsByDepth := range transposed {
+			constraints, err := runtimeInstance.ParachainHostBackingConstraints(paraID)
+			if err != nil {
+				return fmt.Errorf("fetching backing constraints: %w", err)
+			}
+
+			if constraints == nil {
+				logger.Warnf("failed to get inclusion backing constraints")
+				continue
+			}
+
+			pendingAvailabilityCandidates, err := runtimeInstance.
+				ParachainHostCandidatesPendingAvailability(
+					paraID,
+				)
+			if err != nil {
+				return fmt.Errorf("fetching pending availability candidates: %w", err)
+			}
+
+			importablePending, err := pp.preprocessCandidatesPendingAvailability(
+				tmpHeaderCache,
+				constraints,
+				pendingAvailabilityCandidates,
+			)
+			if err != nil {
+				return fmt.Errorf("preprocessing candidates pending availability: %w", err)
+			}
+
+			compactPending := make([]*pendingAvailability, 0, len(importablePending))
+			pendingAvailabilityStorage := newCandidateStorage()
+
+			for _, pending := range importablePending {
+				candidateHash := pending.compact.candidateHash
+
+				err := pendingAvailabilityStorage.addPendingAvailabilityCandidate(
+					candidateHash,
+					pending.candidate,
+					pending.pvd,
+				)
+
+				if err != nil && !errors.Is(err, errCandidateAlreadyKnown) {
+					logger.Warnf("scraped invalid candidate pending availability, "+
+						"candidate hash=%s, para id=%d, err=%s", candidateHash, paraID, err.Error())
+					break
+				}
+
+				compactPending = append(compactPending, &pending.compact)
+			}
+
+			maxBackableChainLen := len(maps.Values(claimsByDepth))
+
+			scope, err := newScopeWithAncestors(
+				relayChainBlockInfo{
+					Hash:        blockInfo.Hash,
+					Number:      blockInfo.Number,
+					StorageRoot: blockInfo.StorageRoot,
+				},
+				constraints,
+				compactPending,
+				uint(maxBackableChainLen),
+				ancestryBlockInfo,
+			)
+			if err != nil {
+				logger.Warnf(
+					"relay chain ancestors have wrong order, "+
+						"para id=%d, max backable=%d, ancestry=%s, leaf=%s, err=%s",
+					paraID,
+					maxBackableChainLen,
+					strings.Join(ancestorsHashes, ","),
+					hash,
+					err.Error(),
+				)
+				continue
+			}
+
+			logger.Tracef(
+				"creating fragment chain, "+
+					"relay parent=%s, min relay parent=%s, max backable=%d, para id=%d, ancestors=%s",
+				hash,
+				scope.earliestRelayParent().Number,
+				maxBackableChainLen,
+				paraID,
+				strings.Join(ancestorsHashes, ","),
+			)
+
+			numOfPendingCandidates := pendingAvailabilityStorage.len()
```

Comment:
can we not just use len(pendingAvailabilityStorage)?

Reply:
pendingAvailabilityStorage is a struct that holds the candidates in a pool, but also some maps for quick search, that's why we use the len method